store/redis_supplier_roles.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package store
+
+import (
+	"context"
+	"fmt"
+
+	l4g "github.com/alecthomas/log4go"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+func (s *RedisSupplier) RoleSave(ctx context.Context, role *model.Role, hints ...LayeredStoreHint) *LayeredStoreSupplierResult {
+	if err := s.client.Del(buildRedisKeyForRoleName(role.Name)).Err(); err != nil {
+		l4g.Error("Redis failed to remove key roles:" + role.Name + " Error: " + err.Error())
+	}
+	return s.Next().RoleSave(ctx, role, hints...)
+}
+
+func (s *RedisSupplier) RoleGet(ctx context.Context, roleId string, hints ...LayeredStoreHint) *LayeredStoreSupplierResult {
+	// Roles are cached by name, as that is most commonly how they are looked up.
+	// This means that no caching is supported on roles being looked up by ID.
+	return s.Next().RoleGet(ctx, roleId, hints...)
+}
+
+func (s *RedisSupplier) RoleGetByName(ctx context.Context, name string, hints ...LayeredStoreHint) *LayeredStoreSupplierResult {
+	var role *model.Role
+	found, err := s.load(buildRedisKeyForRoleName(name), &role)
+	if err != nil {
+		l4g.Error("Redis encountered an error on read: " + err.Error())
+	} else if found {
+		result := NewSupplierResult()
+		result.Data = role
+		return result
+	}
+
+	result := s.Next().RoleGetByName(ctx, name, hints...)
+
+	if result.Err == nil {
+		if err := s.save(buildRedisKeyForRoleName(name), result.Data, REDIS_EXPIRY_TIME); err != nil {
+			l4g.Error("Redis encountered and error on write: " + err.Error())
+		}
+	}
+
+	return result
+}
+
+func (s *RedisSupplier) RoleGetByNames(ctx context.Context, roleNames []string, hints ...LayeredStoreHint) *LayeredStoreSupplierResult {
```

Comment:
Other option is to use the RoleGetByName in this function and make the loop using it. Is sub-optimial, because the SQL query will be repeated for each role name, but is not so common request (I think).

Reply:
Definitely suboptimal on the SQL case.